Check for duplicate layer IDs

ivovandongen opened the pull request:
Fixes #7254

ivovandongen commented:
cc @1ec5 @jfirebaugh

ivovandongen commented on platform/darwin/src/MGLStyle.mm:
```diff
@@ -190,7 +190,11 @@ - (void)addLayer:(MGLStyleLayer *)layer
 
-    self.mapView.mbglMap->addLayer(std::unique_ptr<mbgl::style::Layer>(layer.layer));
+    try {
+        self.mapView.mbglMap->addLayer(std::unique_ptr<mbgl::style::Layer>(layer.layer));
```
Added (tests) for `MGLStyle#insertLayer` and and `MGLMapView#insertCustomStyleLayerWithIdentifier`.

ivovandongen commented on platform/ios/test/MGLLayerTests.m:
```diff
@@ -0,0 +1,25 @@
+#import "MGLMapViewTests.h"
```
Moved the files and added to macos as well. Excluded the custom layer tests for macos as the header is ios specific. Also changed the source test according to https://github.com/mapbox/mapbox-gl-native/pull/7203#discussion_r90384300

1ec5 replied:
That’s fine; the custom style layer API is getting folded into the normal style layer API in #7250, so soon there won’t be anything extra to test.

1ec5 commented on platform/darwin/test/MGLLayerTests.m:
```diff
@@ -0,0 +1,43 @@
+    MGLCustomStyleLayerPreparationHandler preparationHandler = ^{};
+    MGLCustomStyleLayerDrawingHandler drawingHandler = ^(CGSize size, CLLocationCoordinate2D centerCoordinate, double zoomLevel, CLLocationDirection direction, CGFloat pitch, CGFloat perspectiveSkew) {
+    };
+    MGLCustomStyleLayerCompletionHandler completionHander = ^{};
```
FYI, feel free to inline these blocks into the method call. Long lines are par for the course in Objective-C. 😉

1ec5 commented:
The iOS test failure appears to be what I fixed in #7228 on the release-ios-v3.4.0 branch.

ivovandongen commented:
> The iOS test failure appears to be what I fixed in #7228 on the release-ios-v3.4.0 branch.

@1ec5 Could you merge this into master so I can rebase this pr on it?

1ec5 commented on platform/darwin/src/MGLStyle.mm:
```diff
@@ -211,7 +215,11 @@ - (void)insertLayer:(MGLStyleLayer *)layer belowLayer:(MGLStyleLayer *)otherLaye
     const mbgl::optional<std::string> belowLayerId{otherLayer.identifier.UTF8String};
-    self.mapView.mbglMap->addLayer(std::unique_ptr<mbgl::style::Layer>(layer.layer), belowLayerId);
+    try {
+        self.mapView.mbglMap->addLayer(std::unique_ptr<mbgl::style::Layer>(layer.layer), belowLayerId);
```
Now that #7282 has landed, including the changes in #6097, you’ll need to wrap any call to `-[MGLStyleLayer addToMapView:]` in this try-catch block or alternatively move this try-catch block to every implementation of `-addToMapView:` in the concrete subclasses of MGLStyleLayer.

ivovandongen commented:
@1ec5 I've rebased and updated the ios handling/tests: https://github.com/mapbox/mapbox-gl-native/pull/7257/commits/c005962117b3dda2ed663b482625c5ae62fa6333

1ec5 commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -5353,9 +5353,13 @@ - (void)insertCustomStyleLayerWithIdentifier:(NSString *)identifier preparationH
+                                                                      MGLDrawCustomStyleLayer, MGLFinishCustomStyleLayer, context),
+                           otherIdentifier ? mbgl::optional<std::string>(otherIdentifier.UTF8String) : mbgl::optional<std::string>());
+    } catch (std::runtime_error & err) {
+        [NSException raise:@"MGLRedundantLayerIdentiferException" format:@"%s", err.what()];
```
I just realized we’ve been saying “identifer” instead of “identifier”, both for layers and sources. I can fix that typo in a separate PR, since we also need to add a note to the documentation for the affected methods about `MGLRedundantLayerIdentifierException` ([prior art involving sources](https://github.com/mapbox/mapbox-gl-native/blob/93b5ea060237de22b347527f3dc95e6f7cdb7251/platform/darwin/src/MGLStyle.h#L204-L206)).

1ec5 replied:
If you’re in a good spot to do it, can you rename `MGLRedundantLayerIdentiferException` to `MGLRedundantLayerIdentiferException`? I don’t think we have any instances of it in the release branch, so this is a good branch to do it in. I can take care of `MGLRedundantSourceIdentiferException` separately: #7314.

ivovandongen commented:
@kkaefer I've added the backend guard in https://github.com/mapbox/mapbox-gl-native/pull/7257/commits/3f0c6c829443cec27764f025bcd87c6c3121a01b as you suggested and made `activate`/`deactivate` protected.

kkaefer commented on platform/darwin/src/MGLStyle.mm:
```diff
@@ -350,7 +358,11 @@ - (void)addLayer:(MGLStyleLayer *)layer
-    [layer addToMapView:self.mapView];
+    try {
+        [layer addToMapView:self.mapView];
+    } catch (std::runtime_error & err) {
```
We're catching a generic exception here, but always forward a `MGLRedundantLayerIdentifierException`. What we should do long term is to convert all of the generic `std::runtime_error` into individual exception types, and to exception-specific catches that we can then forward.

kkaefer commented on src/mbgl/style/style.cpp:
```diff
@@ -198,6 +198,16 @@ Layer* Style::getLayer(const std::string& id) const {
 Layer* Style::addLayer(std::unique_ptr<Layer> layer, optional<std::string> before) {
     // TODO: verify source
 
+    //Guard against duplicate layer ids
```
Code style is to have a space after `//`.

kkaefer commented on src/mbgl/style/style.cpp:
```diff
@@ -198,6 +198,16 @@ Layer* Style::getLayer(const std::string& id) const {
+
+    if (it != layers.end()) {
+        std::string msg = "Layer " + layer->getID() + " already exists";
+        throw std::runtime_error(msg.c_str());
```
no need to to `c_str()` here; `std::runtime_error` also accepts `std::string`. You could change this to something like this;

```c++
throw std::runtime_error(std::string{"Layer "} + layer->getID() + " already exists");
```

kkaefer commented on test/style/style_layer.test.cpp:
```diff
@@ -268,3 +274,24 @@ TEST(Layer, Observer) {
+        style.addLayer(std::make_unique<LineLayer>("line", "unusedsource"));
+        FAIL() << "Should not have been allowed to add a duplicate layer id";
+    } catch (std::runtime_error) {
+        //Expected
```
Check message here

kkaefer commented on test/style/style_layer.test.cpp:
```diff
@@ -268,3 +274,24 @@ TEST(Layer, Observer) {
+    try {
+        style.addLayer(std::make_unique<LineLayer>("line", "unusedsource"));
+        FAIL() << "Should not have been allowed to add a duplicate layer id";
+    } catch (std::runtime_error) {
```
The rule for exceptions in C++ is to throw by value and catch a `const` reference

kkaefer commented on platform/darwin/src/MGLStyle.mm:
```diff
@@ -271,10 +271,18 @@ - (void)insertObject:(MGLStyleLayer *)styleLayer inLayersAtIndex:(NSUInteger)ind
-        [styleLayer addToMapView:self.mapView];
+        try {
+            [styleLayer addToMapView:self.mapView];
+        } catch (std::runtime_error & err) {
```
`const`

kkaefer commented on platform/darwin/test/MGLStyleTests.mm:
```diff
@@ -165,6 +168,30 @@ - (void)testAddingLayersTwice {
+    XCTAssertThrowsSpecificNamed([self.mapView.style insertLayer:[[MGLFillStyleLayer alloc] initWithIdentifier:@"my-layer" source:source] aboveLayer:initial], NSException, @"MGLRedundantLayerIdentifierException");
+    XCTAssertThrowsSpecificNamed([self.mapView.style insertLayer:[[MGLFillStyleLayer alloc] initWithIdentifier:@"my-layer" source:source] atIndex:0], NSException, @"MGLRedundantLayerIdentifierException");
+
+    #if TARGET_OS_IPHONE
```
Macros are not indented

ivovandongen commented:
@kkaefer Thanks for the review, I've addressed your notes.
